### PR TITLE
Insert Fragment / Drain at End

### PR DESCRIPTION
**Pull request by kprakasam**
Signed-off-by: Prakasam Kannan <kprakasam@me.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

Commit Message:
Currently buffers only provide append, prepend, drain at the beginning of the buffer, this is not sufficient when one modifies (not just append/prepend) the request or response buffer without resorting lot of copying, insert fragment and drain at the end functionality enables such modification while avoiding as much copy as possible.

Additional Description:
We obtain RawSlices underlying request/response buffer and use Ragel based parser to identify certain tags in HTML and inject some content in place this requires moving the data from that point on, copying can be minimized or avoided by inserting a new slice at the modification location, these new methods enable insertion of fragments. 

Risk Level:Low

Testing:
1. Insert tests for SliceDeque
2. Insert tests for Buffer::OwnedImpl

Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by jmarantz** [comment]
Kevin can you take a pass at this? Thanks!

**Comment by kprakasam** [comment]
This InsertAfter/Before signature is kind of weird in taking Slice and it’s index as arguments.  

Most of the streaming parsers require some kind of iterator. I wonder if I should provide a iterator to access ptr to one uint8 of the buffer at a time and a replace function taking the pointer as argument. 

This would reduce the complexity interface and the consuming code, Let me know so I can redo this PR.

File: source/common/buffer/buffer_impl.cc [hunk 1]
```diff
@@ -47,6 +47,48 @@ void OwnedImpl::addBufferFragment(BufferFragment& fragment) {
   slices_.emplace_back(fragment);
 }
 
+std::pair<bool, size_t> findDequeIndex(size_t i, RawSlice& slice, SliceDeque& slices) {
```
Comment (KBaichoo) [inline source/common/buffer/buffer_impl.cc:50]: std::pair seems  unintuitive here, perhaps use optional instead?

File: source/common/buffer/buffer_impl.cc [hunk 2]
```diff
@@ -47,6 +47,48 @@ void OwnedImpl::addBufferFragment(BufferFragment& fragment) {
...
+  size_t empty_slices = 0;
+  size_t deque_index = 0;
+
+  // This is the same logic used by getRawSlices require any changes upstream applied here too
+  for (const auto& deque_slice : slices) {
+    if (deque_slice.dataSize() == 0) {
+      ++empty_slices;
+      continue;
+    }
+    if (deque_index == i && deque_slice.data() == slice.mem_ &&
```
Comment (KBaichoo) [inline source/common/buffer/buffer_impl.cc:60]: What would be the case where `deque_index == i`  it doesn't match the slice given that `getRawSlices()` also uses this logic? In that case why keep iterating and not just return?

Can we get away with using just the RawSlice?

File: include/envoy/buffer/buffer.h [hunk 3]
```diff
@@ -117,6 +117,31 @@ class Instance {
...
+   * done. fragment is not owned. When the fragment->data() is no longer needed, fragment->done() is
+   * called. If RawSlice.mem_ and RawSlice.len_ not same in the Instance, BufferFragment insertion
+   * is not performed and false is returned.
+   *
+   * @param i index of RawSlice in RawSlice array obtained from getRawSlices method.
+   * @param slice slice from RawSlice array obtained from getRawSlices method.
+   * @param fragment the externally owned data to insert in to the buffer.
+   * @return true if insertion was successful or false and no change is made to the buffer
+   */
+  virtual bool insertBufferFragmentAfter(size_t i, RawSlice& slice, BufferFragment& fragment) PURE;
```
Comment (KBaichoo) [inline include/envoy/buffer/buffer.h:131]: Most STL containers that do support insert usually just follow a convention of inserting before given element X, could we possibly do that here as well to narrow the interface? Maybe supporting a stronger iterator is the way to go (i.e. a bidirecitonal iterator)

**Review by KBaichoo — COMMENTED** [review]
I think an iterator might be handy for this -- the multiple insert before / after is  somewhat weird. Perhaps a bidirectional iterator might make sense.

If I grok your original problem correctly you might have some data like `foo BAR foo` -> `foo BAZZ foo`. 

One way we could also implement this without any API changes and minimal copying is just keeping another buffer which we temporarily `move` slices into and then we add our fragment / change and prepend the slices to the original buffer.

File: include/envoy/buffer/buffer.h (hunk 3 above)
Comment (kprakasam) [reply include/envoy/buffer/buffer.h]: Added Bidirectional iterator to buffer. Yet to implement insert that takes and iterator and input.

**Comment by KBaichoo** [comment]
Let me know when this is ready for another pass @kprakasam. Thanks.

**Comment by shivanshu21** [comment]
@kprakasam Are we close to done on this?

**Comment by kprakasam** [comment]
I have to add few more tests will do it this week.

**Pull request closed without merging**